feat(user-service): Include user authorities in JWT (Fixes #5)

HamidShahabi opened the pull request:
This commit addresses issue #5 by adding user authorities to the generated JWT tokens.

- Extracts authorities from the UserDetails object during token generation.
- Adds the authorities as a List<String> to the auth claim in the JWT payload.
- The auth claim will now be used by resource servers for authorization decisions.
- Includes necessary import statements and stream API usage for authority extraction.

HamidShahabi committed:
feat(user-service): Include user authorities in JWT (Fixes #5)

This commit addresses issue #5 by adding user authorities to the generated JWT tokens.

- Extracts authorities from the UserDetails object during token generation.
- Adds the authorities as a List<String> to the auth claim in the JWT payload.
- The auth claim will now be used by resource servers for authorization decisions.
- Includes necessary import statements and stream API usage for authority extraction.

HamidShahabi committed:
feat(user-service): Include user authorities in JWT (Fixes #5)

This commit addresses issue #5 by adding user authorities to the generated JWT tokens.

- Extracts authorities from the UserDetails object during token generation.
- Adds the authorities as a List<String> to the "auth" claim in the JWT payload.
- The "auth" claim will now be used by resource servers for authorization decisions.
- Includes necessary import statements and stream API usage for authority extraction.

HamidShahabi committed:
improvement(user-service): Use lombok annotations instead of redundant getters and setters

HamidShahabi committed:
improvement(user-service): Use lombok annotations instead of redundant getters and setters

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,54 @@
+    private final AuthenticationManager authenticationManager;
+    private final JwtTokenProvider tokenProvider;
+
+    public AuthController(AuthenticationManager authenticationManager, JwtTokenProvider tokenProvider) {
```
lombok RequiredArgConstructor

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,54 @@
+    @PostMapping("/login")
+    public TokenResponse authenticateUser(@RequestBody LoginRequest loginRequest) {
+        try {
+            log.info("Authenticating user with username: {}", loginRequest.getUsername());
```
implement business in a separate layer called service and make controller layer isolated from business

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,54 @@
+        this.tokenProvider = tokenProvider;
+    }
+
+    @PostMapping("/login")
```
consume and produce mediaType

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,54 @@
+            log.error("An unexpected error occurred during authentication", e);
+            throw new RuntimeException("An unexpected error occurred", e);
+        } finally {
+            log.info("Authentication process completed for user: {}", loginRequest.getUsername());
```
delete logs such this
I add a starter to project that it handle logs like this

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,54 @@
+            return new TokenResponse(token);
+        } catch (AuthenticationException e) {
+            log.error("Authentication failed for user: {}", loginRequest.getUsername(), e);
+            throw new RuntimeException("Authentication failed", e);
```
why runtimeException ??
It must throw exception specific for authenticationException 
then implement a RestControllerAdvice that when this exception is thrown then generate error for user with 401 http error code
this RestControllerAdvice must return a ExceptionParamDto object in its method that this dto must contain some param like exceptionName, message and ...

HamidShahabi committed:
improvement(user-service): Use lombok annotations instead of redundant getters and setters

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,54 @@
+            log.error("Authentication failed for user: {}", loginRequest.getUsername(), e);
+            throw new RuntimeException("Authentication failed", e);
+        } catch (Exception e) {
+            log.error("An unexpected error occurred during authentication", e);
```
remove these error level logs
why catch Exception??

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/auth/dto/LoginRequest.java:
```diff
@@ -0,0 +1,20 @@
+    private String username;
+    private String password;
+
+    public LoginRequest() {
```
use NoArgConstructor And AllArgConstructor

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/entity/BaseUser.java:
```diff
@@ -0,0 +1,40 @@
+
+@Data
+@Entity
+public class BaseUser implements Serializable {
```
why BaseUser? for example name it AppUser

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/entity/BaseUser.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.List;
+
+@Data
+@Entity
```
isn't @Table needed?

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/entity/BaseUser.java:
```diff
@@ -0,0 +1,40 @@
+@Entity
+public class BaseUser implements Serializable {
+    @Serial
+    private static final long serialVersionUID = 1L;
```
why use serializable?

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/entity/BaseUser.java:
```diff
@@ -0,0 +1,40 @@
+    private static final long serialVersionUID = 1L;
+    @Id
+    private Long id;
+    private String username;
```
specify type of and attributes of each column with @Column

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/security/config/PasswordEncoderConfig.java:
```diff
@@ -0,0 +1,14 @@
+@Configuration
+public class PasswordEncoderConfig {
+    @Bean
+    public PasswordEncoder passwordEncoder() {
```
move to securityConfig

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/security/config/AuthenticationManagerConfig.java:
```diff
@@ -0,0 +1,44 @@
+ */
+@Configuration
+public class AuthenticationManagerConfig {
+    /**
```
move these bean to securityConfig

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/security/config/JwtTokenProvider.java:
```diff
@@ -0,0 +1,48 @@
+
+    // In a production environment, move this secret to a secure external configuration.
+    private static final String jwtSecret = "fBsWwVrv0BNVpGIRYAjCeXn8nPCp5nHNfMvQY+PI5H9NE37H6ForsPalQRZi01d7zmfLVxe35UIE1qxwWVZxLw==";
+    private static final long jwtExpirationInMs = 3600000; // 1 hour
```
make this configurable

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/repository/UserRepository.java:
```diff
@@ -0,0 +1,11 @@
+public interface UserRepository extends JpaRepository<BaseUser, Long> {
+    BaseUser findByUsername(String username);
+    BaseUser findByEmail(String email);
+    BaseUser findByPhoneNumber(String phoneNumber);
```
delete methods with no usage

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/security/service/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @param userRepository the user repository
+     */
+    public CustomUserDetailsService(UserRepository userRepository) {
```
lombok

HamidShahabi committed:
feat(Auth): Implement service layer and specific exception handling

Refactors authentication logic into a dedicated service.
Adds RestControllerAdvice scoped to AuthController to handle
AuthenticationException (401) and other errors with a standard DTO.
Specifies consumes/produces for login endpoint.

HamidShahabi committed:
feat(config): Externalize and validate JWT configuration

- Makes JWT secret key and expiration time configurable using @Value
  and application.properties (app.jwt.*).
- Uses Lombok @RequiredArgsConstructor for cleaner injection in JwtTokenProvider.
- Adds @PostConstruct validation in JwtTokenProvider to ensure the
  secret is present

refactor(security): Consolidate security configuration beans

- Moves AuthenticationManager, DaoAuthenticationProvider, and
  PasswordEncoder bean definitions from separate config classes
  into the main SecurityConfig class for better cohesion.
- Deletes AuthenticationManagerConfig and PasswordEncoderConfig.
- Updates CSRF configuration syntax to use method reference.

HamidShahabi committed:
chore(db): Remove unused UserRepository methods

- Removes findByEmail and findByPhoneNumber methods from UserRepository
  as they were not used.

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/auth/service/api/AuthenticationService.java:
```diff
@@ -0,0 +1,9 @@
+import com.ftgo.user.userservice.auth.dto.LoginRequest;
+import com.ftgo.user.userservice.auth.dto.TokenResponse;
+
+public interface AuthenticationService {
```
it's too much 
there is no need to have interface for each service

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/exception/handler/RestExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+    public ExceptionParamDto handleAuthenticationException(AuthenticationException ex, HttpServletRequest request) {
+        String path = request.getRequestURI();
+
+        log.warn("Authentication failure at path [{}]: {}", path, ex.getMessage());
```
remove these logs
as i said I add a dependency from tosan github that it handle all  service calls logs

AmirhosseinZDev commented on src/main/java/com/ftgo/user/userservice/exception/handler/RestExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+
+    @ExceptionHandler(Exception.class)
```
generally Exception is for checked exceptions that intend business exception and maybe it must BAD_REQUEST

HamidShahabi committed:
refactor(UserService): use lombok instead of constructor definition.
